hubspot3/test/test_contacts.py

```diff
@@ -1,112 +1,319 @@
+            (20, 20, None),
+            (150, 100, None),  # keep query limit at max 100
+            (10, 10, ["hs_analytics_last_url", "hs_analytics_revenue"]),
+            (10, 10, "lead_source"),
+        ],
+    )
+    def test_get_all(
+        self, contacts_client, mock_connection, limit, query_limit, extra_properties
+    ):
+        response_body = {
+            "contacts": [
+                {
+                    "addedAt": 1390574181854,
+                    "vid": 204727,
+                    "canonical-vid": 204727,
+                    "merged-vids": [],
+                    "properties": {},
+                }
+            ],
+            "has-more": False,
+            "vid-offset": 204727,
+        }
+        get_batch_return = [dict(id=204727 + i) for i in range(30)]
+        get_batch_mock = Mock(return_value=get_batch_return)
+        contacts_client.get_batch = get_batch_mock
+        mock_connection.set_response(200, response_body)
+        resp = contacts_client.get_all(limit=limit, extra_properties=extra_properties)
+        mock_connection.assert_num_requests(1)
+        mock_connection.assert_has_request(
+            "GET",
+            "/contacts/v1/lists/all/contacts/all?count={}&vidOffset=0".format(
+                query_limit
+            ),
+        )
+        assert resp == get_batch_return if not limit else get_batch_return[:limit]
+        get_batch_mock.assert_called_once_with(
+            [contact["vid"] for contact in response_body["contacts"]],
+            extra_properties=extra_properties,
+        )
+
+    @pytest.mark.parametrize(
+        "extra_properties_given, extra_properties_as_list",
+        [
+            (None, []),
+            ("lead_source", ["lead_source"]),
+            (
+                ["hs_analytics_last_url", "hs_analytics_revenue"],
+                ["hs_analytics_last_url", "hs_analytics_revenue"],
+            ),
+        ],
+    )
+    def test_get_batch(
+        self,
+        contacts_client,
+        mock_connection,
+        extra_properties_given,
+        extra_properties_as_list,
+    ):
+        response_body = {
+            "3234574": {"vid": 3234574, "canonical-vid": 3234574, "properties": {}},
+            "3234575": {"vid": 3234575, "canonical-vid": 3234575, "properties": {}},
+        }
+        ids = ["3234574", "3234575"]
+        properties = contacts_client.default_batch_properties.copy()
+        properties.extend(extra_properties_as_list)
+        query_properties = ["property={}".format(p) for p in properties]
+        vids = ["vid={}".format(vid) for vid in ids]
+        params = vids
+        params.extend(query_properties)
+
+        mock_connection.set_response(200, response_body)
+        resp = contacts_client.get_batch(ids, extra_properties_given)
+        mock_connection.assert_num_requests(1)
+        mock_connection.assert_has_request(
+            "GET", "/contacts/v1/contact/vids/batch?{}".format("&".join(params))
```

Comment:
Und auch hier wird sich auf die Reihenfolge eines `dict`s verlassen. Die Utility `assert_query_parameters_in_request` sollte auch hier helfen, die ich in meinem MR nochmal so angepasst habe, dass sie auch mit mehreren Werten pro Parameter zurecht kommt. Da ich zudem auch `assert_has_request` dort nun gefixed habe, sodass es wieder tut, was es soll, könntest du diese Änderungen einfach cherrypicken.